### ISSUE TITLE
[Bug Report] java.lang.NoSuchMethodError: 'com.google.gson.GsonBuilder com.google.gson.GsonBuilder.setObjectToNumberStrategy(com.google.gson.ToNumberStrategy)'

### ISSUE THREAD

**Issue by bnert** [Azure/azure-iot-sdk-java#1656]
# Context

- **OS and version used:** macOS 11.6 (20G165)
- **Java runtime used:**
openjdk version "11.0.16.1" 2022-08-12
OpenJDK Runtime Environment Homebrew (build 11.0.16.1+0)
OpenJDK 64-Bit Server VM Homebrew (build 11.0.16.1+0, mixed mode)
- **SDK version used:** 2.1.3

## Description of the issue
Calling `DeviceClient.getTwin()` results in the below log, followed by the handler timing out.

## Code sample exhibiting the issue
```java
// assume SSLContext sc = proper ssl context setup
ClientOptions co = ClientOptions.builder().sslContext(sc).build();
DeviceClient dc = new DeviceClient(connectionStr, IotHubClientProtocol.MQTT, co);

dc.open();

Twin t = dc.getTwin();
```

## Console log of the issue
```
WARN  c.m.a.s.i.d.t.IotHubTransport  - Updating transport status to new status DISCONNECTED with reason COMMUNICATION_ERROR
com.microsoft.azure.sdk.iot.device.transport.ProtocolException: Mqtt connection lost
        at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException(PahoExceptionTranslator.java:42)
        at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.connectionLost(Mqtt.java:339)
        at org.eclipse.paho.client.mqttv3.internal.CommsCallback.connectionLost(CommsCallback.java:304)
        at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:441)
        at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:228)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: org.eclipse.paho.client.mqttv3.MqttException: MqttException
        ... 2 common frames omitted
Caused by: java.lang.NoSuchMethodError: 'com.google.gson.GsonBuilder com.google.gson.GsonBuilder.setObjectToNumberStrategy(com.google.gson.ToNumberStrategy)'
        at com.microsoft.azure.sdk.iot.device.twin.TwinGsonBuilder.getGson(TwinGsonBuilder.java:27)
        at com.microsoft.azure.sdk.iot.device.twin.Twin.createFromPropertiesJson(Twin.java:317)
        at com.microsoft.azure.sdk.iot.device.twin.DeviceTwin$1.onResponseReceived(DeviceTwin.java:96)
        at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.onMessageReceived(IotHubTransport.java:336)
        at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttIotHubConnection.onMessageArrived(MqttIotHubConnection.java:505)
        at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.messageArrived(Mqtt.java:364)
        at org.eclipse.paho.client.mqttv3.internal.CommsCallback.deliverMessage(CommsCallback.java:519)
        at org.eclipse.paho.client.mqttv3.internal.CommsCallback.handleMessage(CommsCallback.java:417)
        at org.eclipse.paho.client.mqttv3.internal.CommsCallback.run(CommsCallback.java:214)
```

**Comment by bnert** [comment]
This may be due to a conflict w/ another dependency bringing in the gson dependency. Requires further investigation on my end.

**Comment by bnert** [comment]
Looks like this may be due to using `com.microsoft.azure.sdk.iot/iot-service-client@1.18.0`. The library specified brings in `com.google.code.gson/gson@2.5` which is missing the method type specified in the error.

**Comment by timtay-microsoft** [comment]
Let us know if you still see this issue after upgrading to the latest version of the iot-service-client package.

**Comment by bnert** [comment]
@timtay-microsoft, upgraded to to `2.1.3`, and still no luck.

I have made sure to exclude other  libraries from providing `com.google.gson`. This results in `com.google.gson` being provided, however, on reflecting the methods available from the class `setObjectToNumberStrategy` is missing. By excluding `com.google.gson` and providing the latest and `2.9.0`, `setObjectToNumberStrategy` is still missing.

This may be a caching issue on my end.

Something I forgot to mention, is I am using the library from a clojure context.

**Comment by bnert** [comment]
What is strange is the compiled jar's for `com.google.gson@2.10.1` and `com.google.gson@2.9.0` are missing definitions for `setObjectToNumberStrategy`.

The jar files which are being pulled via clojure's leiningen may be pulling the wrong versions. Unzipping the jars which are managed by leiningen them and then run `xxd GsonBuild.class | grep "set"` the function symbols are missing from the class definition, even though the function [present in the source](https://github.com/google/gson/blob/gson-parent-2.9.0/gson/src/main/java/com/google/gson/GsonBuilder.java#L346) from at least `2.9.0`. Pulling the jar's directly from maven result in the function being present.

I'll dig into this on my end some more, maybe leave a comment or two with respect to resolution, but I don't think this is a library issue!

**Comment by timtay-microsoft** [comment]
Okay, thanks for taking a thorough look on your end. Feel free to open a new issue if you still run into issues on this later.

**Comment by bnert** [comment]
Took the following action to resolve:
1. For `lein`, enabled the `:pedantic? :abort` option in `project.clj` in order to make sure `gson` wasn't being included indirectly or munged from another library.
2. Cycled through different version `2.8.9` -> `2.9.0` -> `2.10.1` -> `2.10`, eventually landing on `2.10`. `2.10` was the first version which returned the desired function after inspecting the class methods via `(map #(.getName %) (.getMethods (.getClass (com.google.gson.GsonBuilder.))))` in the repl.

This is most likely some sort of caching/interpretation issue on my local machine, as the desired function is included since `2.8.9`, and the azure library indirectly provides `2.9.0`.
